### PR TITLE
LLM White-box Attack Implementation - Greedy Coordinate Gradient (GCG)

### PR DESCRIPTION
**Pull request by Rockyph**
Key Features
Multi-prompt attack coordination with progressive optimization
Comprehensive conversation template support for various model architectures (LLaMA, GPT, etc.)
Distributed worker processing for scalable attack execution
Robust evaluation framework with jailbreak detection and exact match scoring
Extensive logging and result tracking with JSON output format

Usage
Shell Script
Make executable:
`chmod +x advsecurenet/llm/GCG/experiments/launch_scripts/run_universal_gcg.sh`

Run with default config:
```
cd advsecurenet/llm/GCG/experiments/launch_scripts/
./run_universal_gcg.sh
```

Run with parameters
`./run_universal_gcg.sh "Qwen/Qwen2.5-0.5B-Instruct" "individual" "behaviors" "auto" 100 10 256 0.1`

Arguments: model, attack_type, data_type, device, steps, train_data, batch_size, learning_rate

CLI
`python3 -m advsecurenet.llm.GCG.experiments.cli_gcg attack --config configs/universal_config.py`

Override config parameters:
`python3 cli_gcg.py attack --config configs/universal_config.py --model "gpt2" --steps 50 --device "cpu"`

Configuration
Default settings in experiments/configs/universal_config.py:

Model: Qwen/Qwen2.5-0.5B-Instruct
Attack parameters and data sources configurable
Results auto-saved to [results](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)
Example Notebook
Interactive analysis available in experiments/parse_results.ipynb.

Architecture
Test coverage 95%+ across core modules:

Attack coordination (multi_prompt.py, progressive.py, individual.py, evaluate.py)
Prompt management (attack_prompt.py, prompt_manager.py)
Model integration (model_worker.py, embedding_utils.py)
Conversation handling (template_adapter.py, template_utils.py)
Core GCG algorithm (gcg_attack.py)

### CODE REVIEW

File: advsecurenet/llm/GCG/data/advbench/transfer_expriment_behaviors.csv
Comment (the-anne) [inline advsecurenet/llm/GCG/data/advbench/transfer_expriment_behaviors.csv:1]: Do we need to upload this csv file? Shouldn't this be in gitignore?

File: advsecurenet/llm/GCG/experiments/cli_gcg.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+    steps,
+    train_data,
+    batch_size,
+    learning_rate,
+    control_init,
+    data_offset,
+    verbose,
+    config_override,
+):
+    """Run universal GCG attack on specified model."""
```
Comment (the-anne) [inline advsecurenet/llm/GCG/experiments/cli_gcg.py]: Could you add a more specific docstring, i.e. mention inputs, outputs etc (refer to other files) - Google style

File: advsecurenet/llm/GCG/experiments/configs/universal_config.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    config.attack_type = "individual"  # "individual" or "transfer"
+    config.data_type = "behaviors"  # "behaviors" or "strings"
+    config.attack = "gcg"  # Attack method
+    config.transfer = False
+
+    # === DATA PARAMETERS ===
+    config.n_train_data = 10
+    config.n_test_data = 0
+    config.data_offset = 0
+    config.train_data = "/Users/philip/Desktop/advsecurenet_mp/advsecurenet/llm/GCG/data/advbench/harmful_behaviors.csv"
```
Comment (the-anne) [inline advsecurenet/llm/GCG/experiments/configs/universal_config.py]: you probably want this to be a relative path

File: advsecurenet/llm/GCG/experiments/main.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
+"""A main script to run attack for LLMs."""
```
Comment (the-anne) [inline advsecurenet/llm/GCG/experiments/main.py:1]: Could you add 1-2 sentences more to describe this?

File: advsecurenet/llm/GCG/experiments/results/individual_tiny_llama2_20250831-18:00:07.json
Comment (the-anne) [inline advsecurenet/llm/GCG/experiments/results/individual_tiny_llama2_20250831-18:00:07.json:1]: Do we need to upload this?

File: advsecurenet/llm/GCG/experiments/results/individual_tiny_llama2_20250831-18:08:51.json
Comment (the-anne) [inline advsecurenet/llm/GCG/experiments/results/individual_tiny_llama2_20250831-18:08:51.json:1]: Same question here - applies to every similar json file

Comment (the-anne) [reply advsecurenet/llm/GCG/experiments/results/individual_tiny_llama2_20250831-18:08:51.json:1]: Make sure to remove them

File: advsecurenet/llm/GCG/src/conversation/template_adapter.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+        if not FASTCHAT_AVAILABLE:
+            return None
+
+        # Handle None or empty model name
+        if not model_name:
+            print(f"⚠️ Invalid model name: {model_name}")
+            return None
+
+        try:
+            # First, try direct model name lookup (FastChat's auto-detection)
```
Comment (the-anne) [inline advsecurenet/llm/GCG/src/conversation/template_adapter.py:97]: I think this comment is unnecessary - it is clear what the code does

File: advsecurenet/llm/GCG/src/gcg/__init__.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+# Import only the classes that actually exist in gcg_attack.py
+from .gcg_attack import (
+    GCGAttackPrompt,
+    GCGPromptManager,
+    GCGMultiPromptAttack,
+    token_gradients,
+)
+
+# Import fallbacks for classes that don't exist in GCG
```
Comment (the-anne) [inline advsecurenet/llm/GCG/src/gcg/__init__.py:14]: Unnecessary comment

File: examples/advsecurenet/adversarial_attacks/adversarial_attacks.ipynb
Comment (the-anne) [inline examples/advsecurenet/adversarial_attacks/adversarial_attacks.ipynb:1]: What happened here? Wasn't this notebook already present?

**Review by the-anne — APPROVED** [review]
Great work! I left some comments, mostly about removing redundant files and extending docstrings. Other than that, looks good to me! :D

File: advsecurenet/llm/GCG/src/conversation/template_adapter.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
+import re
+from typing import Dict, Optional, Any, List
+
+try:
```
Comment (fabienmorgan) [inline advsecurenet/llm/GCG/src/conversation/template_adapter.py:18]: What is the reason to put the import into a try/catch block if it doesn't work without it?

File: advsecurenet/llm/GCG/src/prompts/prompt_manager.py [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+                    self.control_str,
+                    best_loss,
+                    runtime,
+                    model_tests,
+                    verbose=verbose,
+                )
+
+                self.control_str = last_control
+
+        return self.control_str, loss, steps
```
Comment (fabienmorgan) [inline advsecurenet/llm/GCG/src/prompts/prompt_manager.py:383]: I don't get the steps variable. Couldn't you just use the n_steps value? Steps gets initialized at 0 and only gets changed in the for loop of range n_steps with an addition of + 1

File: examples/advsecurenet/adversarial_attacks/adversarial_attacks.ipynb
Comment (fabienmorgan) [reply examples/advsecurenet/adversarial_attacks/adversarial_attacks.ipynb:1]: It seems that only the output has changed and that it was accidentally commited without any changes

**Review by fabienmorgan — APPROVED** [review]
Great work
